### ISSUE TITLE
Add OSGi support to JARs

### ISSUE THREAD

**Issue by logoff** [gondor/openstack4j#214]
It would be very useful adding OSGi support in project JARs. Now it is not possible adding project JARs (and dependencies) into and OSGi platform (like [Apache Felix](http://felix.apache.org/) or [Eclipse Equinox](http://eclipse.org/equinox/)).

Adding support to OSGI is very easy using [Apache Felix Maven Bundle Plugin](http://felix.apache.org/site/apache-felix-maven-bundle-plugin-bnd.html).

Finally, creating a [Apache Karaf Feature](http://karaf.apache.org/manual/latest/users-guide/provisioning.html) would ease the job of adding the project as dependency in [Apache Karaf](http://karaf.apache.org/).

Some of the project dependencies are OSGi compatible but there is a mess in their dependency chain (import and export relations are not correct at all!).

**Comment by gondor** [comment]
I would love to have support for OSGI in OpenStack4j.  In my engineering path I've haven't had an environment that leveraged it, so I'm not that versed in it.  Is this something you'd want to contribute?  If not I may have more questions if I look into this.

**Comment by isartcanyameres** [comment]
Hi, I'm currently working on this feature.

The steps I'm following:
- use [Apache Felix Maven Bundle Plugin](https://felix.apache.org/site/apache-felix-maven-bundle-plugin-bnd.html) to create OSGi bundles during project build (it only changes the MANIFEST.MF in generated jar)
- use [Aries SPI Fly](http://aries.apache.org/modules/spi-fly.html#specconf) required headers to make SPI work in OSGi.

I'll rise any question I may have using the [openstack4j group](https://groups.google.com/forum/#!forum/openstack4j)

**Comment by gondor** [comment]
Great, thank you!  I just saw your question in the group.  I'll provide a few answers and as well I can call out who recently did the yaml work so they can respond on those questions.

**Comment by isartcanyameres** [comment]
Although previous patches prepare openstack4j to be depoyed in OSGi environments, currently **openstack4j is not yet functional in OSGi**.

Currently, openstack4j-core cannot be resolved in OSGI as it makes use of a package that is not visible in OSGi environments. 

TODO: 
Find an alternative implementation for org.openstack4j.openstack.heat.utils.Environment.java and org.openstack4j.openstack.heat.utils.Template.java without importing packages that are not visible in OSGi. Please see the [conversation about using not visible package](https://groups.google.com/forum/#!topic/openstack4j/k3W5xn4L7zo)

**Comment by mhshams** [comment]
One more point regarding OSGi support: 
HttpExecutor is using Java ServiceLoader to find out the connector: 
```
Iterator<HttpExecutorService> it = ServiceLoader.load(HttpExecutorService.class).iterator();
```
This solution won't work in OSGi environment, because each bundle ( Jar file in this context) has it's own class loader.

**Comment by isartcanyameres** [comment]
Hi, 
ServiceLoader mechanism is supported in OSGi by using [SPI Fly](http://aries.apache.org/modules/spi-fly.html) dynamic weaving bundle. However, this bundle must be in your OSGi container. 
In case you are using karaf you can make use of the [already defined features](https://github.com/isartcanyameres/openstack4j/blob/e8acf9fc9ee82f8dfd5ebb3f0c2e54382280d327/src/main/resources/features.xml) to install it.

For that to work, bundles must declare (in manifest file) to import/export some capabilities, as explained here: http://aries.apache.org/modules/spi-fly.html#specconf

Openstack4j bundles are now declaring those capabilities in their manifest file as introduced in #242. This is done by means of maven-bundle-plugin configuration in each pom.xml file (i.e. see [httpclient pom.xml](https://github.com/isartcanyameres/openstack4j/commit/b23f68969d0153d6bccfe11b5a86d0132d90903e#diff-bdaaa27f321920744364ee1d56fb43b5L48))

**Comment by timothyjward** [comment]
> ServiceLoader mechanism is supported in OSGi by using SPI Fly dynamic weaving bundle. However, this bundle must be in your OSGi container.

This is true, however it only works for the default usage of `ServiceLoader` (i.e. when you don't pass a `ClassLoader`). This is because the specification works by setting the TCCL to one with visibility of the correct types. The change https://github.com/ContainX/openstack4j/commit/64e1857adea38b008083dd0ba5cef565389bf2f8 (from October 2015) changed this usage, and broke OpenStack4j's OSGi support.

This failure would have been pretty obvious if there were any OSGi integration tests for OpenStack4j. I can recommend the `bnd-testing-maven-plugin` or PAX-Exam. Both would be relatively easy to add to the build, and would prevent regressions like this one.
